src/common.c

```diff
@@ -2951,30 +2951,81 @@ int stlink_erase_flash_page(stlink_t *sl, stm32_addr_t flashaddr) {
   return check_flash_error(sl);
 }
 
+// Check if an address and size are within the flash
+int stlink_check_address_range_validity(stlink_t *sl, stm32_addr_t addr, size_t size) {
+  if (addr < sl->flash_base || addr >= (sl->flash_base + sl->flash_size)) {
+    ELOG("Invalid address, it should be within 0x%08x - 0x%08lx\n", sl->flash_base, (sl->flash_base + sl->flash_size -1));
+    return (-1);
+  }
+  if ((addr + size) > (sl->flash_base + sl->flash_size)) {
+    ELOG("The size exceeds the size of the flash (0x%08lx bytes available)\n", (sl->flash_base + sl->flash_size - addr));
+    return (-1);
+  }
+  return 0;
+}
+
+// Check if an address is aligned with the beginning of a page
+int stlink_check_address_alignment(stlink_t *sl, stm32_addr_t addr) {
+  stm32_addr_t page = sl->flash_base;
+
+  while (page < addr) {
+    page += stlink_calculate_pagesize(sl, page);
+  }
+
+  if (page != addr) {
+    return -1;
+  }
+
+  return 0;
+}
+
+int stlink_erase_flash_section(stlink_t *sl, stm32_addr_t base_addr, size_t size, bool align_size) {
+  // Check the address and size validity
+  if (stlink_check_address_range_validity(sl, base_addr, size) < 0) {
+    return -1;
+  }
+
+  // Make sure the requested address is aligned with the beginning of a page
+  if (stlink_check_address_alignment(sl, base_addr) < 0) {
+    ELOG("The address to erase is not aligned with the beginning of a page\n");
+    return -1;
+  }
+
+  stm32_addr_t addr = base_addr;
+  do {
+    size_t page_size = stlink_calculate_pagesize(sl, addr);
+
+    // Check if size is aligned with a page, unless we want to completely erase the last page
+    if ((addr + page_size) > (base_addr + size) && !align_size) {
+      ELOG("Invalid size (not aligned with a page). Page size at address %#x is %#lx\n", addr, page_size);
+      return -1;
+    }
+
+    if (stlink_erase_flash_page(sl, addr)) {
+      WLOG("Failed to erase_flash_page(%#x) == -1\n", addr);
+      return (-1);
+    }
+
+    fprintf(stdout, "-> Flash page at %#x erased (size: %#lx)\n", addr, page_size);
```

Comment:
Please, add explicit conversation to `unsigned long` for `page_size`

Reply:
see comment below